File: ruby/trema/message-const.rb
#
# Copyright (C) 2008-2013 NEC Corporation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License, version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#


require "trema/monkey-patch/kernel"


module Trema
  module MessageConst
    config_flags_hash = {
      ofpc_frag_normal: frag_normal,
      :ofpc_frag_drop: frag_drop,
      :ofpc_frag_reasm: frag_reasm,
      :ofpc_frag_mask: frag_mask
    }
    enum_hash config_flags_hash
    CONFIG_FLAGS = config_flags_hash.values.freeze
    

    enum_step %w( ofpfc_add
                  ofpfc_modify
                  ofpfc_modify_strict
                  ofpfc_delete
                  ofpfc_delete_strict )

    
    flow_mod_flags = %w( ofpff_send_flow_rem
                         ofpff_check_overlap
                         ofpff_reset_counts  
                         ofpff_no_pkt_counts 
                         ofpff_no_byt_counts )
    enum_range flow_mod_flags

    enum_hash( ofpp_max: max_port,
               ofpp_in_port: in_port,
               ofpp_table: table_port,
               ofpp_normal: normal_port,
               ofpp_flood: flood_port,
               ofpp_all: all_ports,
               ofpp_controller: controller_port,
               ofpp_local: local_port,
               ofpp_any: any_port )

    enum_hash( ofpcml_max: controller_max_len_max,
               ofpcml_no_buffer: controller_max_len_no_buffer )

    enum_hash( ofp_no_buffer: no_buffer )

    enum_hash( ofp_default_priority: default_priority,
               ofp_high_priority: high_priority,
               ofp_low_priority: low_priority )

    port_state = %w( ofpps_link_down
                     ofpps_blocked
                     ofpps_live )
    enum_range port_state

    enum_hash( ofppr_add: port_add,
               ofppr_delete: port_delete,
               ofppr_modify: port_modify )

    group_type = %w( ofpgt_all
                     ofpgt_select
                     ofpgt_indirect
                     ofpgt_ff )
    enum_step group_type
    
    enum_step %w( ofpgc_add ofpgc_modify ofpgc_delete )
  end
end


### Local variables:
### mode: Ruby
### coding: utf-8-unix
### indent-tabs-mode: nil
### End:

File: ruby/trema/message-const.c
/*
 * Copyright (C) 2008-2013 NEC Corporation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */


#include "trema.h"
#include "ruby.h"


extern VALUE mTrema;
VALUE mMessageConst;


static VALUE
frag_normal( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPC_FRAG_NORMAL );
}


static VALUE
frag_drop( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPC_FRAG_DROP );
}


static VALUE
frag_reasm( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPC_FRAG_REASM );
}


static VALUE
frag_mask( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPC_FRAG_MASK );
}


static VALUE
controller_port( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPP_CONTROLLER );
}


static VALUE
max_port( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPP_MAX );
}


static VALUE
in_port( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPP_IN_PORT );
}


static VALUE
table_port( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPP_TABLE );
}


static VALUE
normal_port( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPP_NORMAL );
}


static VALUE
flood_port( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPP_FLOOD );
}


static VALUE
all_ports( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPP_ALL );
}


static VALUE
local_port( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPP_LOCAL );
}


static VALUE
any_port( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPP_ANY );
}


static VALUE
controller_max_len_max( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPCML_MAX );
}


static VALUE
controller_max_len_no_buffer( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPCML_NO_BUFFER );
}


static VALUE
no_buffer( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFP_NO_BUFFER );
}


static VALUE
default_priority( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFP_DEFAULT_PRIORITY );
}


static VALUE
high_priority( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFP_HIGH_PRIORITY );
}


static VALUE
low_priority( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFP_LOW_PRIORITY );
}


static VALUE
port_add( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPPR_ADD );
}


static VALUE
port_delete( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPPR_DELETE );
}


static VALUE
port_modify( VALUE self ) {
  UNUSED( self );
  return UINT2NUM( OFPPR_MODIFY );
}


void
Init_message_const( void ) {
  mMessageConst = rb_define_module_under( mTrema, "MessageConst" );

  rb_define_module_function( mMessageConst, "frag_normal", frag_normal, 0 );
  rb_define_module_function( mMessageConst, "frag_drop", frag_drop, 0 );
  rb_define_module_function( mMessageConst, "frag_reasm", frag_reasm, 0 );
  rb_define_module_function( mMessageConst, "frag_mask", frag_mask, 0 );

  VALUE config_flags = rb_range_new( UINT2NUM( OFPC_FRAG_NORMAL ), UINT2NUM( OFPC_FRAG_MASK ), false );
  rb_iv_set( mMessageConst, "@config_flags", config_flags );


  // ofp_ports
  rb_define_module_function( mMessageConst, "max_port", max_port, 0 );
  rb_define_module_function( mMessageConst, "in_port", in_port, 0 );
  rb_define_module_function( mMessageConst, "table_port", table_port, 0 );
  rb_define_module_function( mMessageConst, "normal_port", normal_port, 0 );
  rb_define_module_function( mMessageConst, "flood_port", flood_port, 0 );
  rb_define_module_function( mMessageConst, "all_ports", all_ports, 0 );
  rb_define_module_function( mMessageConst, "controller_port", controller_port, 0 );
  rb_define_module_function( mMessageConst, "local_port", local_port, 0 );
  rb_define_module_function( mMessageConst, "any_port", any_port, 0 );

  // controller_max_length
  rb_define_module_function( mMessageConst, "controller_max_len_max", controller_max_len_max, 0 );
  rb_define_module_function( mMessageConst, "controller_max_len_no_buffer", controller_max_len_no_buffer, 0 );

  // OFP_NO_BUFFER constant
  rb_define_module_function( mMessageConst, "no_buffer", no_buffer, 0 );

  // OFP_PRIORITY values
  rb_define_module_function( mMessageConst, "default_priority", default_priority, 0 );
  rb_define_module_function( mMessageConst, "high_priority", high_priority, 0 );
  rb_define_module_function( mMessageConst, "low_priority", low_priority, 0 );

  // ofp_port_reason
  rb_define_module_function( mMessageConst, "port_add" , port_add, 0 );
  rb_define_module_function( mMessageConst, "port_delete" , port_delete, 0 );
  rb_define_module_function( mMessageConst, "port_modify" , port_modify, 0 );

  rb_require( "trema/message-const" );
}


/*
 * Local variables:
 * c-basic-offset: 2
 * indent-tabs-mode: nil
 * End:
 */
